fvm/src/kernel/default.rs

```diff
@@ -864,6 +873,94 @@ where
             .create_actor(code_id, actor_id, delegated_address)
     }
 
+    fn upgrade_actor<K: Kernel>(
+        &mut self,
+        new_code_cid: Cid,
+        params_id: BlockId,
+    ) -> Result<SendResult> {
+        if self.read_only {
+            return Err(
+                syscall_error!(ReadOnly, "upgrade_actor cannot be called while read-only").into(),
+            );
+        }
+
+        // Load parameters.
+        let params = if params_id == NO_DATA_BLOCK_ID {
+            None
+        } else {
+            Some(self.blocks.get(params_id)?.clone())
+        };
+
+        // Make sure we can actually store the return block.
+        if self.blocks.is_full() {
+            return Err(syscall_error!(LimitExceeded; "cannot store return block").into());
+        }
+
+        let result = self.call_manager.with_transaction(|cm| {
+            let origin = cm.origin();
+
+            let state = cm
+                .get_actor(self.actor_id)?
+                .ok_or_else(|| syscall_error!(NotFound; "actor not found"))?;
+
+            // store the code cid of the calling actor before running the upgrade entrypoint
+            // in case it was changed (which could happen if the target upgrade entrypoint
+            // sent a message to this actor which in turn called upgrade)
+            let code = state.code;
+
+            // update the code cid of the actor to new_code_cid
+            cm.set_actor(
+                origin,
+                ActorState::new(
+                    new_code_cid,
+                    state.state,
+                    state.balance,
+                    state.sequence,
+                    None,
+                ),
+            )?;
+
+            // run the upgrade entrypoint
+            let result = cm.call_actor::<Self>(
+                self.caller,
+                Address::new_id(self.actor_id),
+                Entrypoint::Upgrade(UpgradeInfo { old_code_cid: code }),
+                params,
+                &TokenAmount::from_whole(0),
+                None,
+                false,
+            )?;
+
+            if result.exit_code == ExitCode::OK {
+                // after running the upgrade, our code cid must not have changed
+                let code_after_upgrade = cm
+                    .get_actor(self.actor_id)?
+                    .ok_or_else(|| syscall_error!(NotFound; "actor not found"))?
+                    .code;
+                if code != code_after_upgrade {
+                    return Err(syscall_error!(Forbidden; "re-entrant upgrade detected").into());
+                }
```

Comment:
I _think_ this check should actually be in `send`, not `upgrade`. I.e.:

1. If an actor calls `A (upgrade -> upgrade -> upgrade)`, that's fine. Because, on success, we'll "unwind" past _all_ the upgrade calls all at once.
2. If an actor has a call stack like `A -> B -> A (upgrade)`, that's _not_ fine and we need to abort when we return to the top-level call into `A`. Otherwise, we'd end up running old code.

It's that last point that's problematic: We never want to re-enter some actor code after we've upgraded away from it.

Reply:
And, in that second case, we should treat it as a "revert" of `B` (I think?).

Reply:
Actually, after talking this through with @maciejwitowski,  there's an alternative here that I kind of like. Instead of catching this when we return to the top-level invocation of `A`, we can catch this on the upgrade syscall itself. To do this, we'd need to:

1. Keep a map of `ActorID -> reentrancy count`, incrementing the count each time we re-enter an actor already on the call-stack.
2. In the upgrade syscall, if this count is non-zero (for the current actor), return a `Forbidden` syscall error (or something like that).

One reason I like this approach is that, independent of _this_ particular change, I'd like to find a way to _expose_ whether or not a call is reentrant to actors. I.e., some kind of flag in their environment where they can detect that they're within a reentrant call. I expect most actors will, at that point, simply abort with an error.

Reply:
@Stebalien I played around with this approach in [109c5e9](https://github.com/filecoin-project/ref-fvm/pull/1866/commits/109c5e9fd7f8369e9ef3d3f940a432c8e276ee10)

Reply:
Ah, I mean that we need to keep track of _calls_ in progress, not upgrades. I.e., if some actor `A` is already on the call stack, no "deeper" instance of `A` should be able to call the `upgrade` syscall.

Reply:
Oh lets talk sync today

Reply:
We can remove this now, right?